File: core/redactor.go
package core

import (
	"regexp"
	"strings"

	"github.com/cyrildever/feistel"
	"github.com/cyrildever/feistel/common/utils/base256"
	"github.com/cyrildever/redacted/model"
)

//--- TYPES

// Redactor ...
type Redactor struct {
	Dictionary model.Dictionary
	Tag        string
	Cipher     *feistel.FPECipher
}

//--- METHODS

// Redact ...
func (r Redactor) Redact(line string, delimiters ...string) string {
	var actualDelimiters []string
	for _, delim := range delimiters {
		if delim != " " && delim != "\t" && delim != "\n" {
			actualDelimiters = append(actualDelimiters, delim)
		}
	}
	if len(actualDelimiters) == 0 {
		actualDelimiters = append(actualDelimiters, "\\s")
	}
	var outputs []string
	re := regexp.MustCompile(strings.Join(actualDelimiters, ""))
	words := re.Split(line, -1)
	for _, word := range words {
		if r.Dictionary.Contains(word) {
			obfuscated, err := r.Cipher.Encrypt(word)
			if err != nil {
				obfuscated = base256.ToBase256Readable([]byte(word))
			}
			outputs = append(outputs, obfuscated.String())
		} else {
			outputs = append(outputs, word)
		}
	}
	return strings.Join(outputs, " ")
}

// Expand ...
func (r Redactor) Expand(line string, delimiters ...string) string {
	var actualDelimiters []string
	for _, delim := range delimiters {
		if delim != " " && delim != "\t" && delim != "\n" {
			actualDelimiters = append(actualDelimiters, delim)
		}
	}
	if len(actualDelimiters) == 0 {
		actualDelimiters = append(actualDelimiters, "\\s")
	}
	var outputs []string
	re := regexp.MustCompile(strings.Join(actualDelimiters, ""))
	words := re.Split(line, -1)
	for _, word := range words {
		if deciphered, err := r.Cipher.Decrypt(base256.Readable(word)); err == nil {
			if r.Dictionary.Contains(deciphered) {
				outputs = append(outputs, deciphered)
			} else {
				outputs = append(outputs, word)
			}
		} else {
			outputs = append(outputs, word)
		}
	}
	return strings.Join(outputs, " ")
}

//--- FUNCTIONS

// NewRedactor ...
func NewRedactor(dictionary model.Dictionary, tag string, cipher *feistel.FPECipher) (r *Redactor) {
	if dictionary.NonEmpty() {
		if tag != "" {
			r = &Redactor{
				Dictionary: dictionary,
				Tag:        tag,
				Cipher:     cipher,
			}
		} else {
			r = &Redactor{
				Dictionary: dictionary,
				Cipher:     cipher,
			}
		}
	} else {
		r = &Redactor{
			Tag:    tag,
			Cipher: cipher,
		}
	}
	return
}

// NewRedactorWithDictionary ...
func NewRedactorWithDictionary(dic model.Dictionary, cipher *feistel.FPECipher) *Redactor {
	return NewRedactor(dic, "", cipher)
}

// NewRedactorWithDictionary ...
func NewRedactorWithTag(tag string, cipher *feistel.FPECipher) *Redactor {
	return NewRedactor(model.Dictionary{}, tag, cipher)
}

File: scala/src/main/scala/fr/edgewhere/redacted/core/Redactor.scala
package fr.edgewhere.redacted.core

import fr.edgewhere.redacted.model._
import fr.edgewhere.feistel.FormatPreservingEncryption
import scala.util.control.Breaks._

import Redactor._

/**
 * Redactor ...
 *
 * @author  Cyril Dever
 * @since   1.0
 * @version 1.0
 */
final case class Redactor(dictionary: Dictionary, tag: Tag, cipher: FormatPreservingEncryption, both: Boolean = true) {
  self =>

  def redact(line: String, delimiters: String*): String = {
    line.split(("\\s" ++ delimiters.filter(d => d != "\\s" && d != " " && d != "\t" && d != "\n")).mkString).map(word => {
      var output = word
      breakable {
        if (self.both || self.dictionary.nonEmpty) {
          if (self.dictionary.contains(word)) {
            output = self.cipher.encrypt(word)
            break
          } else if (self.tag.nonEmpty && word.startsWith(self.tag)) {
            output = self.tag + self.cipher.encrypt(word.substring(self.tag.length))
            break
          }
        } else if (!self.both && self.dictionary.isEmpty) {
          if (self.tag.nonEmpty && word.startsWith(self.tag)) {
            output = self.tag + self.cipher.encrypt(word.substring(self.tag.length))
            break
          }
        }
      }
      output
    }).mkString(" ")
  }

  def expand(line: String): String =
    line.split(" ").map(word => {
      var output = word
      breakable {
        if (self.both || self.dictionary.nonEmpty) {
          if (self.tag.nonEmpty && word.startsWith(self.tag)) {
            output = self.tag + self.cipher.decrypt(word.substring(self.tag.length))
            break
          } else if (dictionary.contains(self.cipher.decrypt(word))) {
            output = self.cipher.decrypt(word)
            break
          }
        } else if (!self.both && self.dictionary.isEmpty) {
          if (self.tag.nonEmpty && word.startsWith(self.tag)) {
            output = self.tag + self.cipher.decrypt(word.substring(self.tag.length))
            break
          }
        }
      }
      output
    }).mkString(" ")

  def clean(str: String): String = if (self.tag.nonEmpty) str.replace(self.tag, "") else str
}
object Redactor {
  type Tag = String

  def apply(dictionary: Dictionary, cipher: FormatPreservingEncryption): Redactor = Redactor(dictionary, "", cipher, false)

  def apply(tag: Tag, cipher: FormatPreservingEncryption): Redactor = Redactor(Dictionary.EMPTY, tag, cipher, false)
}
